File: backend/common/cbor.py
"""
Deterministic CBOR codec — byte-identical to Kotlin core-crypto/CborCodec.

Rules (RFC 8949 deterministic):
  - Map keys sorted by lexicographic byte-wise comparison of their UTF-8 bytes.
    (All our maps have string keys, so this is equivalent to sorted(key.encode('utf-8'))).
  - Integers use minimal-width encoding (RFC 8949 §4.2.1).
  - No indefinite lengths; all arrays/maps have definite lengths.
  - No floats; only unsigned ints, negative ints, byte strings, text strings, arrays, maps, null, bool.
  - Byte strings are major type 2; text strings major type 3; both length-prefixed.

Implementation sits on top of `cbor2` but enforces determinism explicitly by:
  - Sorting map keys before encoding.
  - Using cbor2's canonical mode for integer width.

Alternatively a hand-rolled encoder is included (encode_* helpers) so behavior is
fully specified without depending on cbor2 internals. The public `encode_*` functions
use the hand-rolled encoder; `decode` uses cbor2 for convenience.

Golden vectors in golden_vectors.json are produced by this encoder and must match
Kotlin's output byte-for-byte. The hex strings there are the ground truth.
"""
from __future__ import annotations

import struct
from typing import Any, Dict, List, Tuple

try:
    import cbor2  # type: ignore
    _HAS_CBOR2 = True
except Exception:
    _HAS_CBOR2 = False


# ---------------------------------------------------------------------------
# Low-level CBOR primitives (hand-rolled, deterministic)
# ---------------------------------------------------------------------------

def _encode_uint(major: int, value: int) -> bytes:
    """Encode unsigned int with major type, minimal width."""
    assert value >= 0
    mt = major << 5
    if value < 24:
        return bytes([mt | value])
    elif value < 256:
        return bytes([mt | 24, value])
    elif value < 65536:
        return bytes([mt | 25]) + struct.pack(">H", value)
    elif value < 4294967296:
        return bytes([mt | 26]) + struct.pack(">I", value)
    else:
        return bytes([mt | 27]) + struct.pack(">Q", value)


def _encode_int(n: int) -> bytes:
    if n >= 0:
        return _encode_uint(0, n)
    else:
        # CBOR negative: value = -1 - n
        return _encode_uint(1, -1 - n)


def _encode_bytes(b: bytes) -> bytes:
    return _encode_uint(2, len(b)) + b


def _encode_text(s: str) -> bytes:
    raw = s.encode("utf-8")
    return _encode_uint(3, len(raw)) + raw


def _encode_array(items: List[bytes]) -> bytes:
    head = _encode_uint(4, len(items))
    return head + b"".join(items)


def _encode_map(pairs: List[Tuple[bytes, bytes]]) -> bytes:
    """
    pairs: list of (encoded_key_bytes, encoded_value_bytes) where encoded_key
    is the CBOR encoding of the key (always a text string in our schema).
    Sorted by raw key bytes (the UTF-8 of the key) per spec — but since keys are
    text strings, sorting by encoded key bytes == sorting by UTF-8 bytes with
    length prefix tie-break which for our short ASCII keys is equivalent to
    lexicographic on the string. We sort by the *encoded key* bytes to be precise.
    """
    # sort by encoded key bytes
    pairs_sorted = sorted(pairs, key=lambda kv: kv[0])
    head = _encode_uint(5, len(pairs_sorted))
    out = head
    for k_enc, v_enc in pairs_sorted:
        out += k_enc + v_enc
    return out


def _encode_any(value: Any) -> bytes:
    """Encode Python value to deterministic CBOR."""
    if value is None:
        return bytes([0xF6])  # null
    if isinstance(value, bool):
        return bytes([0xF5 if value else 0xF4])
    if isinstance(value, int):
        return _encode_int(value)
    if isinstance(value, bytes):
        return _encode_bytes(value)
    if isinstance(value, str):
        return _encode_text(value)
    if isinstance(value, list):
        return _encode_array([_encode_any(x) for x in value])
    if isinstance(value, dict):
        pairs: List[Tuple[bytes, bytes]] = []
        for k, v in value.items():
            if not isinstance(k, str):
                raise TypeError(f"CBOR map keys must be str, got {type(k)}")
            k_enc = _encode_text(k)
            v_enc = _encode_any(v)
            pairs.append((k_enc, v_enc))
        return _encode_map(pairs)
    raise TypeError(f"unsupported CBOR type: {type(value)}")


def encode_canonical(obj: Any) -> bytes:
    """
    Encode any JSON-like structure to deterministic CBOR.
    This is the single entry point for golden-vector generation.
    """
    return _encode_any(obj)


def decode_canonical(data: bytes) -> Any:
    """Decode CBOR bytes (uses cbor2 if available, else raises)."""
    if not _HAS_CBOR2:
        raise RuntimeError("cbor2 not installed; cannot decode")
    return cbor2.loads(data)


# ---------------------------------------------------------------------------
# Domain-specific encoders — canonical maps for signed structures
# Each returns the CBOR bytes that are signed (signature field excluded).
# Key ordering is enforced by _encode_map sorting, but we list keys explicitly
# for readability and to ensure Kotlin uses the same set.
# ---------------------------------------------------------------------------

def encode_manifest_signing(manifest) -> bytes:
    """
    Manifest signing payload:
    {
      "blobId": bytes(32),
      "chunks": [bytes(32), ...],
      "expiresAt": int|null,
      "mimeType": text,
      "publishedAt": int,
      "publisherId": bytes(32),
      "totalBytes": int
    }
    Keys sorted: blobId, chunks, expiresAt, mimeType, publishedAt, publisherId, totalBytes
    """
    m: Dict[str, Any] = {
        "blobId": manifest.blob_id.merkle_root,
        "chunks": [c.sha256 for c in manifest.chunks],
        "expiresAt": manifest.expires_at,  # None -> CBOR null
        "mimeType": manifest.mime_type,
        "publishedAt": manifest.published_at,
        "publisherId": manifest.publisher_id.public_key,
        "totalBytes": manifest.total_bytes,
    }
    return encode_canonical(m)


def encode_manifest_full(manifest) -> bytes:
    """Full manifest including signature (for storage / transport, not for signing)."""
    m: Dict[str, Any] = {
        "blobId": manifest.blob_id.merkle_root,
        "chunks": [c.sha256 for c in manifest.chunks],
        "expiresAt": manifest.expires_at,
        "mimeType": manifest.mime_type,
        "publishedAt": manifest.published_at,
        "publisherId": manifest.publisher_id.public_key,
        "signature": manifest.signature,
        "totalBytes": manifest.total_bytes,
    }
    return encode_canonical(m)


def encode_receipt_signing(receipt) -> bytes:
    """
    DeliveryReceipt signing payload:
    {
      "blobId": bytes(32),
      "bytesDelivered": int,
      "nonce": bytes(16),
      "recipientId": bytes(32),
      "senderId": bytes(32),
      "timestamp": int
    }
    Sorted: blobId, bytesDelivered, nonce, recipientId, senderId, timestamp
    """
    m: Dict[str, Any] = {
        "blobId": receipt.blob_id.merkle_root,
        "bytesDelivered": receipt.bytes_delivered,
        "nonce": receipt.nonce,
        "recipientId": receipt.recipient_id.public_key,
        "senderId": receipt.sender_id.public_key,
        "timestamp": receipt.timestamp,
    }
    return encode_canonical(m)


def encode_contribution_signing(contrib) -> bytes:
    """
    Contribution signing payload:
    {
      "capturedAt": int,
      "contributorId": bytes(32),
      "correctedText": text,
      "id": bytes(16)  -- UUID as 16-byte big-endian
      "modelVersion": text|null,
      "originalModelOutput": text|null,
      "sourceLang": text,
      "sourceText": text,
      "targetLang": text
    }
    Sorted: capturedAt, contributorId, correctedText, id, modelVersion, originalModelOutput, sourceLang, sourceText, targetLang
    """
    m: Dict[str, Any] = {
        "capturedAt": contrib.captured_at,
        "contributorId": contrib.contributor_id.public_key,
        "correctedText": contrib.corrected_text,
        "id": contrib.id.bytes,  # 16 bytes big-endian
        "modelVersion": contrib.model_version,
        "originalModelOutput": contrib.original_model_output,
        "sourceLang": contrib.source_lang,
        "sourceText": contrib.source_text,
        "targetLang": contrib.target_lang,
    }
    return encode_canonical(m)


def encode_card_tx_signing(tx) -> bytes:
    """
    CardTxRecord signing payload:
    {
      "amount": int,
      "balanceAfter": int,
      "cardId": bytes,
      "counter": int,
      "nonce": bytes,
      "timestamp": int
    }
    Sorted: amount, balanceAfter, cardId, counter, nonce, timestamp
    """
    m: Dict[str, Any] = {
        "amount": tx.amount,
        "balanceAfter": tx.balance_after,
        "cardId": tx.card_id,
        "counter": tx.counter,
        "nonce": tx.nonce,
        "timestamp": tx.timestamp,
    }
    return encode_canonical(m)


# ---------------------------------------------------------------------------
# Utility
# ---------------------------------------------------------------------------

def to_hex(b: bytes) -> str:
    return b.hex()


def from_hex(s: str) -> bytes:
    return bytes.fromhex(s)

File: android/core-crypto/src/main/java/net/sharenet/crypto/Cbor.kt
package net.sharenet.crypto

import java.io.ByteArrayOutputStream

/**
 * Deterministic (canonical) CBOR encoder/decoder.
 *
 * Implements the subset needed for ShareNet signed structures:
 *  - unsigned int  (major 0)
 *  - negative int  (major 1)
 *  - byte string   (major 2)
 *  - text string   (major 3, UTF-8)
 *  - array         (major 4)
 *  - map           (major 5) — keys sorted lexicographically by UTF-8 bytes
 *  - simple values (major 7): false (0xF4), true (0xF5), null (0xF6)
 *
 * Canonical rules (RFC 7049 §3.9 / RFC 8949 §4.2):
 *  - Integers use the shortest encoding.
 *  - Map keys are sorted by their encoded byte representation; for this
 *    implementation keys are always text strings, so sorting by UTF-8
 *    string value is equivalent and deterministic.
 *  - No indefinite-length items.
 *  - Duplicate map keys are forbidden.
 *
 * This codec is shared with the Python backend via golden-vector tests.
 * Any change must be reflected in `backend/common/cbor.py` and re-validated
 * against `golden-vectors.json`.
 */
object Cbor {

    // -----------------------------------------------------------------------
    // Public high-level encoders — these define the canonical signing form.
    // -----------------------------------------------------------------------

    /**
     * Encode a [Manifest] for signing. The `signature` field is excluded;
     * the signer signs exactly the bytes returned here.
     */
    fun encodeManifest(m: Manifest): ByteArray = encode(
        sortedMapOf(
            "blobId" to m.blobId.bytes,
            "chunks" to m.chunks.map { it.bytes },
            "expiresAt" to m.expiresAt,
            "mimeType" to m.mimeType,
            "publishedAt" to m.publishedAt,
            "publisherId" to m.publisherId.bytes,
            "totalBytes" to m.totalBytes,
        ),
    )

    /**
     * Encode a [DeliveryReceipt] for signing. The `recipientSignature` field is excluded.
     */
    fun encodeReceipt(r: DeliveryReceipt): ByteArray = encode(
        sortedMapOf(
            "blobId" to r.blobId.bytes,
            "bytesDelivered" to r.bytesDelivered,
            "nonce" to r.nonce,
            "recipientId" to r.recipientId.bytes,
            "senderId" to r.senderId.bytes,
            "timestamp" to r.timestamp,
        ),
    )

    /**
     * Encode a [Contribution] for signing. The `signature` field is excluded.
     */
    fun encodeContribution(c: Contribution): ByteArray = encode(
        sortedMapOf(
            "capturedAt" to c.capturedAt,
            "contributorId" to c.contributorId.bytes,
            "correctedText" to c.correctedText,
            "id" to c.id.toString(),
            "modelVersion" to c.modelVersion,
            "originalModelOutput" to c.originalModelOutput,
            "sourceLang" to c.sourceLang,
            "sourceText" to c.sourceText,
            "targetLang" to c.targetLang,
        ),
    )

    // -----------------------------------------------------------------------
    // Generic encode / decode
    // -----------------------------------------------------------------------

    /**
     * Encode an arbitrary value to canonical CBOR bytes.
     *
     * Supported types:
     *  - `null`
     *  - `Boolean`
     *  - `Int`, `Long`
     *  - `ByteArray` → byte string
     *  - `String` → text string
     *  - `List<*>` → array
     *  - `Map<String, *>` → map with sorted keys
     *
     * Other types throw [IllegalArgumentException].
     */
    fun encode(value: Any?): ByteArray {
        val out = ByteArrayOutputStream()
        encodeValue(value, out)
        return out.toByteArray()
    }

    /**
     * Decode CBOR bytes to a Kotlin value. Inverse of [encode].
     * Returns `null`, `Boolean`, `Long`, `ByteArray`, `String`, `List<Any?>`, or `Map<String, Any?>`.
     */
    fun decode(bytes: ByteArray): Any? {
        val decoder = Decoder(bytes)
        val result = decoder.decodeValue()
        require(decoder.isAtEnd()) { "Trailing bytes after CBOR value" }
        return result
    }

    // -----------------------------------------------------------------------
    // Internal encoder
    // -----------------------------------------------------------------------

    private fun encodeValue(value: Any?, out: ByteArrayOutputStream) {
        when (value) {
            null -> out.write(0xF6)
            is Boolean -> out.write(if (value) 0xF5 else 0xF4)
            is Int -> encodeInt(value.toLong(), out)
            is Long -> encodeInt(value, out)
            is ByteArray -> {
                encodeHead(2, value.size.toLong(), out)
                out.write(value)
            }
            is String -> {
                val utf8 = value.toByteArray(Charsets.UTF_8)
                encodeHead(3, utf8.size.toLong(), out)
                out.write(utf8)
            }
            is List<*> -> {
                encodeHead(4, value.size.toLong(), out)
                for (elem in value) encodeValue(elem, out)
            }
            is Map<*, *> -> {
                // Canonical: sort keys by string representation lexicographically.
                @Suppress("UNCHECKED_CAST")
                val map = value as Map<String, Any?>
                val sortedKeys = map.keys.sorted()
                encodeHead(5, sortedKeys.size.toLong(), out)
                for (k in sortedKeys) {
                    encodeValue(k, out)
                    encodeValue(map[k], out)
                }
            }
            else -> throw IllegalArgumentException("Unsupported CBOR type: ${value::class.java.name}")
        }
    }

    private fun encodeInt(v: Long, out: ByteArrayOutputStream) {
        if (v >= 0) {
            encodeHead(0, v, out)
        } else {
            // CBOR negative: value -1 - n
            encodeHead(1, -1 - v, out)
        }
    }

    private fun encodeHead(major: Int, value: Long, out: ByteArrayOutputStream) {
        val mt = major shl 5
        when {
            value < 24 -> out.write(mt or value.toInt())
            value < 256 -> {
                out.write(mt or 24)
                out.write(value.toInt())
            }
            value < 65536 -> {
                out.write(mt or 25)
                out.write((value ushr 8).toInt() and 0xFF)
                out.write(value.toInt() and 0xFF)
            }
            value < 4294967296L -> {
                out.write(mt or 26)
                out.write((value ushr 24).toInt() and 0xFF)
                out.write((value ushr 16).toInt() and 0xFF)
                out.write((value ushr 8).toInt() and 0xFF)
                out.write(value.toInt() and 0xFF)
            }
            else -> {
                out.write(mt or 27)
                for (shift in 56 downTo 0 step 8) {
                    out.write(((value ushr shift) and 0xFF).toInt())
                }
            }
        }
    }

    // -----------------------------------------------------------------------
    // Internal decoder
    // -----------------------------------------------------------------------

    private class Decoder(private val bytes: ByteArray) {
        var pos: Int = 0

        fun isAtEnd(): Boolean = pos == bytes.size

        fun decodeValue(): Any? {
            require(pos < bytes.size) { "Unexpected end of CBOR data" }
            val initial = bytes[pos++].toInt() and 0xFF
            val major = initial ushr 5
            val info = initial and 0x1F
            return when (major) {
                0 -> readUInt(info)
                1 -> -1 - readUInt(info)
                2 -> {
                    val len = readUInt(info).toInt()
                    val slice = bytes.copyOfRange(pos, pos + len)
                    pos += len
                    slice
                }
                3 -> {
                    val len = readUInt(info).toInt()
                    val slice = bytes.copyOfRange(pos, pos + len)
                    pos += len
                    String(slice, Charsets.UTF_8)
                }
                4 -> {
                    val len = readUInt(info).toInt()
                    val list = ArrayList<Any?>(len)
                    repeat(len) { list.add(decodeValue()) }
                    list
                }
                5 -> {
                    val len = readUInt(info).toInt()
                    val map = LinkedHashMap<String, Any?>(len)
                    repeat(len) {
                        val k = decodeValue()
                        require(k is String) { "Map key must be text string, got ${k?.let { it::class.java.name }}" }
                        map[k] = decodeValue()
                    }
                    map
                }
                7 -> when (info) {
                    20 -> false
                    21 -> true
                    22 -> null
                    else -> throw IllegalArgumentException("Unsupported simple value: $info")
                }
                else -> throw IllegalArgumentException("Unsupported CBOR major type: $major")
            }
        }

        private fun readUInt(info: Int): Long = when (info) {
            in 0..23 -> info.toLong()
            24 -> (bytes[pos++].toInt() and 0xFF).toLong()
            25 -> {
                val v = ((bytes[pos].toInt() and 0xFF) shl 8) or (bytes[pos + 1].toInt() and 0xFF)
                pos += 2
                v.toLong()
            }
            26 -> {
                var v = 0L
                repeat(4) { v = (v shl 8) or (bytes[pos++].toInt() and 0xFF).toLong() }
                v
            }
            27 -> {
                var v = 0L
                repeat(8) { v = (v shl 8) or (bytes[pos++].toInt() and 0xFF).toLong() }
                v
            }
            else -> throw IllegalArgumentException("Invalid additional info: $info")
        }
    }
}

/**
 * Helper to build a [LinkedHashMap] with insertion order = sorted key order.
 * Useful for deterministic encoding where call-site map literal order must not matter.
 */
private fun sortedMapOf(vararg pairs: Pair<String, Any?>): Map<String, Any?> {
    val sorted = pairs.sortedBy { it.first }
    val map = LinkedHashMap<String, Any?>(sorted.size)
    for ((k, v) in sorted) map[k] = v
    return map
}
